### ISSUE TITLE
Touching custom contentView will close the calloutView

### ISSUE THREAD

**Issue by guilhermearaujo** [nfarina/calloutview#62]
Using the "standard" view (with title + subtitle), if I tap the callout view, nothing will happen (or `calloutViewClicked` will be called, if the delegate is implemented).
However if I set a custom `contentView` to my `calloutView` and then tap that custom view, the `calloutView` won't handle the touch. Instead, the map will dismiss it.

How can I avoid this behavior and prevent the callout view from hiding?

**Comment by akhatmullin** [comment]
I'm try add breakpoint at first line of method 'calloutClicked' in file SMCalloutView.m and this has no effect. This method not calling when I tap calloutView.

**Comment by nfarina** [comment]
Hello, what kind of view is the callout being presented in? Are you using a `MKMapView`?

**Comment by akhatmullin** [comment]
Hello!
I'm using not only MKMapView. GMSMapView using too.

**Comment by akhatmullin** [comment]
Methods <code>highlightIfNecessary</code> and <code>unhighlightIfNecessary</code>

**Comment by nfarina** [comment]
Yes the problem here is that, in those map controls, touch handling is controlled strictly by the map view itself. So you need to do some work to work around that. Check the README for a short discussion on Google's map view, and the Sample project should contain code demoing how to work with MKMapView and receive touches.

**Comment by akhatmullin** [comment]
Yes, and we having mistake exactly into Sample project.
This is from file <code>MapKitComparisonController.m (line 115)</code>
<code>[self.calloutView presentCalloutFromRect:annotationView.bounds inView:annotationView constrainedToView:self.view animated:YES];</code>
If we will to present callout as in this example delegate's methods not working.

I using this code into <code>mapView:didSelectAnnotationView:</code>:
<code>
CLLocationCoordinate2D coordinate = [mapView convertPoint:annotationView.center toCoordinateFromView:mapView];
CGPoint point = [mapView convertCoordinate:coordinate toPointToView:mapView];
    
CGRect calloutRect = CGRectZero;
calloutRect.origin = point;
calloutRect.size = CGSizeZero;
    
calloutView.calloutOffset = CGPointMake(0, -annotationView.frame.size.height/2);
</code>

**Comment by hiroara** [comment]
Hello. I probably also faced same issue, but it works for me by overriding `hitTest:point:withEvent` of `MKAnnotationView`.

I leave writing code but do not know what would be helpful.


```
// written by swift
override func hitTest(point: CGPoint, withEvent event: UIEvent?) -> UIView? {
    return self.subviews.reduce(nil) { hitView, subview in
        if hitView != nil { return hitView }
        let pointForTargetView = subview.convertPoint(point, fromView: self)

        if CGRectContainsPoint(subview.bounds, pointForTargetView) {
            return subview.hitTest(pointForTargetView, withEvent: event)
        } else {
            return nil
        }

    } ?? super.hitTest(point, withEvent: event)
}

```

I referred to [this article](https://developer.apple.com/library/ios/qa/qa2013/qa1812.html).

**Comment by lelandrichardson** [comment]
FYI, I ran into this issue and was able to fix the problem by tweaking the following line in the overridden `gestureRecognizer:shouldReceiveTouch` method on [this line](https://github.com/nfarina/calloutview/blob/master/SampleProject/MapKitComparisonController.m#L180):

Essentially:
```
if ([touch.view isKindOfClass:[UIControl class]])
```
becomes
```
if ([touch.view isDescendantOfView:self.calloutView])
```

**Comment by jpm** [comment]
Looks like every year we need another update on this ticket, so here's mine. :)

I also reproduced this issue on my app. To be explicit, we are doing a custom `contentView` and the `calloutViewClicked:` delegate method would never fire. The custom `contentView` has a bunch of labels in it, plus a very small `UIButton`. Tapping anywhere on the callout would just dismiss it, and the delegate method never fired.

What I did was expose the internal `containerView` property as readonly:

```objc
@property (nonatomic, strong, readonly) UIButton *containerView; // for masking and interaction
```

Then I modified the example `MKMapView` subclass:

```objc
// override UIGestureRecognizer's delegate method so we can prevent MKMapView's recognizer from firing
// when we interact with UIControl subclasses inside our callout view.
- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch
{
    if ([touch.view isKindOfClass:[UIControl class]])
        return NO;
    else
        return [super gestureRecognizer:gestureRecognizer shouldReceiveTouch:touch];
}

// Allow touches to be sent to our calloutview.
// See this for some discussion of why we need to override this: https://github.com/nfarina/calloutview/pull/9
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{
    UIView *calloutMaybe = [self.calloutView hitTest:[self.calloutView convertPoint:point fromView:self] withEvent:event];
    if ([calloutMaybe isKindOfClass:[UIControl class]]) {
        return calloutMaybe;
    } else if (calloutMaybe) {
        return self.calloutView.containerView;
    }
    
    return [super hitTest:point withEvent:event];
}
```

This way if the hit test returns a UIControl subclass, then we pass the event on to it, otherwise we fall back to the main `UIButton` object which handles the callout taps (aka `containerView`).

@nfarina Let me know if you would like me to provide a PR for it.
